streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -924,22 +924,64 @@ private StreamThread createStreamThread(final long cacheSizePerThread, final int
      * @return name of the added stream thread or empty if a new stream thread could not be added
      */
     public Optional<String> addStreamThread() {
-        synchronized (changeThreadCount) {
-            if (isRunningOrRebalancing()) {
-                final int threadIdx = getNextThreadIndex();
-                final long cacheSizePerThread = getCacheSizePerThread(threads.size() + 1);
+        if (isRunningOrRebalancing()) {
+            final int threadIdx;
+            final long cacheSizePerThread;
+            synchronized (changeThreadCount) {
+                threadIdx = getNextThreadIndex();
+                cacheSizePerThread = getCacheSizePerThread(threads.size() + 1);
                 resizeThreadCache(cacheSizePerThread);
-                final StreamThread streamThread = createStreamThread(cacheSizePerThread, threadIdx);
-                synchronized (stateLock) {
-                    if (isRunningOrRebalancing()) {
-                        streamThread.start();
-                        return Optional.of(streamThread.getName());
-                    } else {
-                        streamThread.shutdown();
+            }
+            final StreamThread streamThread = createStreamThread(cacheSizePerThread, threadIdx);
+
+            synchronized (stateLock) {
+                if (isRunningOrRebalancing()) {
+                    streamThread.start();
+                    return Optional.of(streamThread.getName());
+                } else {
+                    streamThread.shutdown();
+                    threads.remove(streamThread);
+                    resizeThreadCache(getCacheSizePerThread(threads.size()));
+                    return Optional.empty();
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Removes one stream thread out of the running stream threads from this Kafka Streams client.
+     * <p>
+     * The removed stream thread is gracefully shut down. This method does not specify which stream
+     * thread is shut down.
+     * <p>
+     * Since the number of stream threads decreases, the sizes of the caches in the remaining stream
+     * threads are adapted so that the sum of the cache sizes over all stream threads equals the total
+     * cache size specified in configuration {@link StreamsConfig#CACHE_MAX_BYTES_BUFFERING_CONFIG}.
+     *
+     * @return name of the removed stream thread or empty if a stream thread could not be removed because
+     *         no stream threads are alive
+     */
+    public Optional<String> removeStreamThread() {
+        if (isRunningOrRebalancing()) {
+            for (final StreamThread streamThread : threads) {
```

Comment:
We need to protect this with a lock or use a thread-safe data structure for `threads`, otherwise we can get a ConcurrentModificationException if the user calls addThread and/or removeThread at the same time (on that note let's add test coverage for this)

Reply:
Where is it synchronized? I'm thinking that eg while iterating through this list in one thread, another thread may be inside the `synchronized (changeThreadCount)` block of where it removes from this list. Only the second access is synchronized so the iteration itself is still vulnerable right?

Reply:
https://github.com/apache/kafka/blob/338045c9b62c3b6154789d9cfa9a1cf104188e16/streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java#L825

The list access is synchronized. I think that when using an iterator such as a for each loop that should be safe. But since we are not actually making changes out side of the `synchronized (changeThreadCount)` those shouldn't cause a problem.

Reply:
Ah I missed the previous PR where you changed the list type. Ok I'm  not that familiar with `Collections.synchronizedList` but I'm still worried we may not be safe with this. From the [javadocs](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedList(java.util.List)):

```
It is imperative that the user manually synchronize on the returned list when iterating over it:
```

Reply:
It wouldn't be guaranteed to catch this, but either way I think we should have a test that starts up two threads which both try to `removeThread()` at the same time (and maybe similarly for `addStreamThread`)

Reply:
Okay we won't get a concurrent modification but the result won't be deterministic, I think we can just synchronize the iterator in remove thread as that is the only time were we are iterating and making changes. The other places concurrent access should not be a problem. 

I've added tests for the remove and add, there does not seem to be a problem but it is hard to guarantee that the race condition hit each time.